From: Martin Burnicki
Subject:ntpd doesn't synchronize to local clock (ntpd 4.2.8p4/xenial)
Bug: http://bugs.ntp.org/show_bug.cgi?id=2965#c0
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/ntp/+bug/1558125

Martin Burnicki 2015-11-16 10:28:57 UTC

ntpd 4.2.8p4 doesn't synchronize to local clock anymore. This still used to
work in 4.2.8p3.

It turned out that this new bug was introduced by the changeset for bug 2625:
http://bugs.ntp.org/show_bug.cgi?id=2625#c5

where flag1 support by the local clock was removed.

In the original code pp->leap was set to either LEAP_ADDSECOND or to
LEAP_NOWARNING, depending on CLK_FLAG1. Now this has been removed completely,
and pp->leap isn't set at all.

Thus the local clock always claims to be unsynchronized, and thus it is never
accepted as system peer.

diff --git a/ntpd/refclock_local.c b/ntpd/refclock_local.c
index d816c55..8c0f74f 100644
--- a/ntpd/refclock_local.c
+++ b/ntpd/refclock_local.c
@@ -205,6 +205,7 @@ local_poll(
 	pp->disp = 0;
 	pp->jitter = 0;
 #else /* KERNEL_PLL LOCKCLOCK */
+	pp->leap = LEAP_NOWARNING;
 	pp->disp = DISPERSION;
 	pp->jitter = 0;
 #endif /* KERNEL_PLL LOCKCLOCK */
